src/wp-includes/wp-db.php

```diff
@@ -53,6 +53,15 @@
  */
 class wpdb {
 
+	/**
+	 * Version number, to check what features are available.
+	 *
+	 * @since 6.0.0 - V2: wbdb::prepare() supports Identifiers via 'SELECT * FROM %i', and Variadics via 'IN (%...d)'
+	 *
+	 * @var int
+	 */
+	public $version = 2;
```

Comment:
Has a deliberate decisions been made to have `$wpdb->version` start at a low number in the 6.0 release? That's fine if intentional, but it might surprise someone to see such a low integer version number for a class that has existed for 15+ years.

Also: how will plugin developers make use of this in practice? I assume something like this, if they want to maintain back compat with older WP versions:

```php
if ( isset( $wpdb->version ) && $wpdb->version > 1 ) {
    $sql = $wpdb->prepare( "SELECT * FROM %i", $mytable );
} else {
    $sql = "SELECT * FROM `" . esc_sql( $mytable ) . "`"; // or something equally old-fashioned
}
```

I have no better ideas, just thinking through how it's likely to be used.

Reply:
This was after a discussion with [John Blackbourn](https://core.trac.wordpress.org/ticket/54042#comment:13), and you're right, it's only there so plugins can check if they can use `%i`, and provide a fallback if necessary. John did suggest "a new wpdb method", but I'm not sure about that - would that be `prepare2()`, `prepareIdentifiers()`, or something else?

I prefer to have all db escaping in one place, so I went with a simple version number to support this (to be used exactly like your example). I'm open to suggestions, even if you would like to start with a different version number, maybe it could be 6 to match 6.0? or for a bit of fun, the good old 42?

Reply:
Is there any reason this couldn't just tick the `$wp_db_version` or even use `version_compare` with `$wp_version`?

Reply:
I believe John was thinking about how anyone could have their own `/wp-content/db.php` file, which might not support this feature (i.e. providing their own `prepare()` method), and that would stay in place as they upgrade the rest of their WordPress install (why it cannot be `$wp_version`). As to `$wp_db_version`, it has a similar problem, and I believe that's more for the database schema, rather than features wpdb provides.

Reply:
I initially suggested a new method for the sake of simplicity but I agree that this isn't ideal because 15+ years of documentation points to `$wpdb->prepare()`.

That said, a `wpdb->version` property suffers from the same problem as checking `$wp_version` or `$wp_db_version` -- a custom db.php implementation which overrides the `prepare` method with one which does not support the `%i` placeholder still inherits the `version = 2` property from the `wpdb` parent class.

Another option is to introduce a method which runs a smoke test. Example:

```php
public function supports_i() {
	return ( '%i' !== $this->prepare( '%i', 'hello' ) );
}
```

Usage:

```php
if ( method_exists( $wpdb, 'supports_i' ) && $wpdb->supports_i() ) {
    $sql = $wpdb->prepare( "SELECT * FROM %i", $mytable );
} else {
    $sql = "SELECT * FROM `" . esc_sql( $mytable ) . "`"; // or something equally old-fashioned
}
```

Thoughts?

Reply:
> Another option is to introduce a method which runs a smoke test. Example:

This may be a terrible idea, but could `wpdb::has_cap()` be used for this? 

Currently the method is used exclusively for database server features, so this would be require expanding its purpose. 

Feature testing started with the `supports_collation()` method but it was deprecated in favour of a general purpose `has_cap()`. I guess to avoid the need for `method_exists()` type code.

Reply:
> This may be a terrible idea, but could `wpdb::has_cap()` be used for this?

That seems like a good fit. I'd favour using a clearly readable name:

```php
if ( $wpdb->has_cap( 'identifier_placeholders' ) ) {
  $sql = $wpdb->prepare( "SELECT * FROM %i", $mytable );
} else {
  $sql = "SELECT * FROM `" . esc_sql( $mytable ) . "`"; // or something equally old-fashioned
}
```

Having a clear, standardized idiom would help facilitate wpcs support and eventual deprecation of the old code.

I'm starting to wonder if we should explicitly recommend the old `%1$s` syntax for this:

```php
if ( $wpdb->has_cap( 'identifier_placeholders' ) ) {
  $sql = $wpdb->prepare( "SELECT * FROM %i", $mytable );
} else {
  $sql = $wpdb->prepare( "SELECT * FROM `%1$s`", $mytable );
}
```

That way wpcs will be happier, and we could in future have `prepare()` trigger a deprecation warning on the old syntax in future. I think that's better than recommending roll-your-own `esc_sql()` idioms that are open-ended. It gives plugin devs a way to write code now that is clean, and both backwards and forwards compatible.

Reply:
Using `wpdb::has_cap()` seems perfect to me for this and #2191 (even more so). I imagine this would only be used for  3-4 WordPress versions before plugins started just requiring `WordPress 6.0+`.

It suffers the same problem with database dropins though, but not as badly.. A dropin may have replaced `has_cap()` and it might not pass unknown value tests onto the parent `wpdb` class. But in those cases, at least the calling code will go into "legacy mode".
I can't see any valid reason why one would override `prepare()` but not `has_cap()`, so I can't see an issue with that.

It's been very common in the past for plugins to branch based on `$wp_version`, `if ( version_compare( $wp_version, '6.0', '>=' ) { ... new style ... } else { ... old style ... }` when they couldn't do a `function_exists()`, but I can't think of any time we'd want to encourage that coding style, which is why introducing `$wpdb->version` doesn't make sense to me.

Reply:
Agreed, +1 for using `has_cap()` 👍

Reply:
Thanks, I like that approach as well. I’m out walking today/tomorrow, so will implement on Wednesday.

Reply:
I've added support for `$wpdb->has_cap( 'identifier_placeholders' )` and removed `$wpdb->version`.

I've also replaced the `@return int|false` for `has_cap()` on [line 3864](https://github.com/WordPress/wordpress-develop/blob/1b8c0dde173f28bfd2ccd68f3203edda04a36ccd/src/wp-includes/wp-db.php#L3864), as all of the other feature checks use `version_compare()` with "the optional `operator` argument, [where] the function will return true if the relationship is the one specified by the operator, false otherwise." ([source](https://www.php.net/version_compare))